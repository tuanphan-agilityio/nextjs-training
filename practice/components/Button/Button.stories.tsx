import { Meta, Story } from '@storybook/react';

import Button, { ButtonProps } from '.';
import FacebookIcon from '../../public/icons/facebook.svg';

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    variant: {
      control: {
        type: 'select',
        options: ['primary', 'secondary', 'tertiary'],
      },
    },
  },
} as Meta;

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  variant: 'primary',
  children: 'Primary Button',
};

export const Secondary = Template.bind({});
Secondary.args = {
  variant: 'secondary',
  children: 'Secondary Button',
  startIcon: <FacebookIcon />,
};

export const Tertiary = Template.bind({});
Tertiary.args = {
  variant: 'tertiary',
  children: 'Tertiary Button',
};
