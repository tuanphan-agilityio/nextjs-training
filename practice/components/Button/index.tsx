import { ButtonHTMLAttributes, FC, ReactElement, memo } from 'react';
import clsx from 'clsx';

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: 'primary' | 'secondary' | 'tertiary';
  startIcon?: ReactElement;
  endIcon?: ReactElement;
  className?: string;
}

const buttonVariants = {
  primary:
    'bg-gradient-to-b from-primary-gradient-start to-primary-gradient-end text-primary',
  secondary: 'text-secondary',
  tertiary:
    'text-transparent font-primary-bold bg-clip-text bg-gradient-to-b from-primary-gradient-start to-primary-gradient-end text-primary',
};

const Button: FC<ButtonProps> = ({
  variant = 'primary',
  children = '',
  className = '',
  startIcon = null,
  endIcon = null,
  ...props
}) => {
  const buttonClasses = clsx(
    'inline-flex justify-center items-center w-[180px] h-12 bg-primary font-primary-regular text-md hover:opacity-80',
    buttonVariants[variant],
    className,
  );

  return (
    <button className={buttonClasses} {...props}>
      {startIcon && <span className='inline-block mr-5'>{startIcon}</span>}
      {children}
      {endIcon && <span className='inline-block ml-5'>{endIcon}</span>}
    </button>
  );
};

export type { ButtonProps };

export default memo(Button);
