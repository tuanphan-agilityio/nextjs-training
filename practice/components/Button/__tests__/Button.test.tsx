import { fireEvent, render, screen } from '@testing-library/react';
import Button, { ButtonProps } from '..';

describe('Button Component', () => {
  const handleClick = jest.fn();

  const mockProps: ButtonProps = {
    children: 'Add To Cart',
    onClick: handleClick,
  };

  it('Should show match Button DOM Snapshot', () => {
    const { container } = render(<Button {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  it('Should call onClick when the button is clicked', () => {
    render(<Button {...mockProps} />);

    const button = screen.getByRole('button');

    fireEvent.click(button);

    expect(handleClick).toBeCalled();
    expect(handleClick).toBeCalledTimes(1);
  });

  it('Should have the secondary variant class when variant is "secondary"', () => {
    render(<Button {...mockProps} variant='secondary' />);
    const button = screen.getByRole('button');
    expect(button).toHaveClass('text-secondary');
  });
});
