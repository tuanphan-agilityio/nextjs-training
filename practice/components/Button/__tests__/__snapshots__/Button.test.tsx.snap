// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Should show match Button DOM Snapshot 1`] = `
<div>
  <button
    class="inline-flex justify-center items-center w-[180px] h-12 bg-primary font-primary-regular text-md hover:opacity-80 bg-gradient-to-b from-primary-gradient-start to-primary-gradient-end text-primary button-class"
  >
    <span
      class="inline-block mr-5"
    >
      <span>
        startIcon
      </span>
    </span>
    Add To Cart
    <span
      class="inline-block ml-5"
    >
      <span>
        endIcon
      </span>
    </span>
  </button>
</div>
`;
